infrastructure/tari_util/src/chacha20.rs

```diff
@@ -0,0 +1,269 @@
+fn construct_state(key: &[u8; 32], nonce: &[u32; 3], counter: u32) -> [u32; 16] {
+    let constant: [u8; 16] = [101, 120, 112, 97, 110, 100, 32, 51, 50, 45, 98, 121, 116, 101, 32, 107]; // 0x61707865, 0x3320646e, 0x79622d32, 0x6b206574
+    let mut state_bytes = constant.to_vec(); // 128 bit
+    state_bytes.extend_from_slice(key); // 256-bit
+    state_bytes.extend_from_slice(&u32_to_u8_le(counter)); // 32-bit
+    state_bytes.extend_from_slice(&u32_to_u8_le(nonce[0])); // 96-bit
+    state_bytes.extend_from_slice(&u32_to_u8_le(nonce[1]));
+    state_bytes.extend_from_slice(&u32_to_u8_le(nonce[2]));
+    // Convert [u8;64] to [u32;16]
+    let mut curr_bytes: [u8; 4] = [0; 4];
+    let mut state: [u32; 16] = [0; 16];
+    for i in 0..state.len() {
+        let byte_index = i * 4;
+        curr_bytes.copy_from_slice(&state_bytes[byte_index..(byte_index + 4)]);
+        state[i] = u8_to_u32_le(&curr_bytes);
+    }
+    (state)
+}
+
+/// Generate a keystream consisting of a number of chacha blocks
+fn chacha20_cipher_keystream(key: &[u8; 32], nonce: &[u32; 3], block_count: usize) -> Vec<u8> {
+    let bytes_per_block: usize = 64;
+    let block_byte_count = block_count * bytes_per_block;
+    let mut cipher_bytes: Vec<u8> = Vec::with_capacity(block_byte_count as usize);
+    for counter in 1..=block_count as u32 {
+        let state = construct_state(key, nonce, counter);
+        let cipher_block = chacha20_block(&state);
+        // convert cipher block to bytes
+        let mut block_bytes: Vec<u8> = Vec::with_capacity(bytes_per_block);
+        for i in 0..cipher_block.len() {
+            block_bytes.append(&mut u32_to_u8_le(cipher_block[i]).to_vec())
+        }
+        cipher_bytes.append(&mut block_bytes)
+    }
+    (cipher_bytes)
+}
+
+/// Encode the provided input bytes using a chacha20 keystream and a default nonce
+pub fn encode(bytes: &Vec<u8>, key: &[u8; 32]) -> Vec<u8> {
+    let default_nonce: [u32; 3] = [0; 3];
+    (encode_with_nonce(bytes, key, &default_nonce))
+}
+
+/// Encode the provided input bytes using a chacha20 keystream
+pub fn encode_with_nonce(bytes: &Vec<u8>, key: &[u8; 32], nonce: &[u32; 3]) -> Vec<u8> {
+    let bytes_per_block: usize = 64;
+    let block_count = (bytes.len() as f64 / bytes_per_block as f64).ceil() as usize;
+    let cipher_bytes = chacha20_cipher_keystream(key, nonce, block_count);
+    let mut encoded_bytes = Vec::with_capacity(bytes.len());
+    for i in 0..bytes.len() {
+        encoded_bytes.push(cipher_bytes[i] ^ bytes[i]);
+    }
+    (encoded_bytes)
+}
+
+/// Decode the provided input bytes using a chacha20 keystream and a default nonce
+pub fn decode(bytes: &Vec<u8>, key: &[u8; 32]) -> Vec<u8> {
+    (encode(bytes, &key))
+}
+
+/// Decode the provided input bytes using a chacha20 keystream
+pub fn decode_with_nonce(bytes: &Vec<u8>, key: &[u8; 32], nonce: &[u32; 3]) -> Vec<u8> {
+    (encode_with_nonce(bytes, &key, &nonce))
```

Comment:
nit: I'm surprised this doesn't return as a tuple of length 1 (Is that the same thing as a "single" value in rust?). Any reason for the brackets around the returns in your functions or is that just a style choice?

Reply:
I'm not sure, for me it is a style choice to mark the return value.